fundingmanager_test.go

```diff
@@ -0,0 +1,596 @@
+var bobFundingMgr *fundingManager
+
+var miningNode *rpctest.Harness
+
+var aliceTestDir string
+var bobTestDir string
+
+func setupFundingManagers(t *testing.T) {
+	// We need to set the global config, as fundingManager uses
+	// MaxPendingChannels, and it is usually set in lndMain()
+	cfg = &config{
+		MaxPendingChannels: defaultMaxPendingChannels,
+	}
+
+	netParams := &chaincfg.SimNetParams
+
+	// Initialize the harness around a btcd node which will serve as our
+	// dedicated miner to generate blocks, cause re-orgs, etc. We'll set
+	// up this node with a chain length of 125, so we have plentyyy of BTC
+	// to play around with.
+	var err error
+	miningNode, err = rpctest.New(netParams, nil, nil)
+	if err != nil {
+		t.Fatalf("unable to create mining node: %v", err)
+	}
+	if err = miningNode.SetUp(true, 25); err != nil {
+		t.Fatalf("unable to set up mining node: %v", err)
+	}
+
+	// Next mine enough blocks in order for segwit and the CSV package
+	// soft-fork to activate on SimNet.
+	numBlocks := netParams.MinerConfirmationWindow * 2
+	if _, err = miningNode.Node.Generate(numBlocks); err != nil {
+		t.Fatalf("unable to generate blocks: %v", err)
+	}
+
+	rpcConfig := miningNode.RPCConfig()
+
+	chainNotifier, err := btcdnotify.New(&rpcConfig)
+	if err != nil {
+		t.Fatalf("unable to create notifier: %v", err)
+	}
+	if err = chainNotifier.Start(); err != nil {
+		t.Fatalf("unable to start notifier: %v", err)
+	}
+	aliceTestDir, err = ioutil.TempDir("", "alicelnwallet")
+	if err != nil {
+		t.Fatalf("unable to create temp directory: %v", err)
+	}
+
+	aliceMsgChan = make(chan lnwire.Message)
+	aliceAnnounceChan = make(chan lnwire.Message)
+	aliceFundingMgr, err = createTestFundingManager(t, alicePubKey, aliceTestDir, alicePrivKeyBytes[:], netParams, rpcConfig, miningNode, chainNotifier, aliceMsgChan, aliceAnnounceChan)
+	if err != nil {
+		t.Fatalf("failed creating fundingManager: %v", err)
+	}
+	if err = aliceFundingMgr.Start(); err != nil {
+		t.Fatalf("failed starting fundingManager: %v", err)
+	}
+
+	bobTestDir, err = ioutil.TempDir("", "boblnwallet")
+	if err != nil {
+		t.Fatalf("unable to create temp directory: %v", err)
+	}
+	bobMsgChan = make(chan lnwire.Message)
+	bobAnnounceChan = make(chan lnwire.Message)
+	bobFundingMgr, err = createTestFundingManager(t, bobPubKey, bobTestDir, bobPrivKeyBytes[:], netParams, rpcConfig, miningNode, chainNotifier, bobMsgChan, bobAnnounceChan)
+	if err != nil {
+		t.Fatalf("failed creating fundingManager: %v", err)
+	}
+	if err = bobFundingMgr.Start(); err != nil {
+		t.Fatalf("failed starting fundingManager: %v", err)
+	}
+}
+
+func tearDownFundingManagers() {
+	os.RemoveAll(bobTestDir)
+	os.RemoveAll(aliceTestDir)
+	miningNode.TearDown()
+}
+
+func TestFundingManager_normalWorkflow(t *testing.T) {
+	setupFundingManagers(t)
+	defer tearDownFundingManagers()
+
+	// Create a funding request and start the workflow
+	errChan := make(chan error, 1)
+	updateChan := make(chan *lnrpc.OpenStatusUpdate, 1)
```

Comment:
At a quick glance, the buffer of this needs to be _two_ otherwise it'll be possible for one of the goroutines to get suck indefinitely on a send.